tinygrad/runtime/support/hcq.py

```diff
@@ -1,9 +1,28 @@
 from __future__ import annotations
 from typing import List, Optional, Dict, Tuple, cast, Protocol, Type, Union, TypeVar, Generic, Callable, ParamSpec, Concatenate
-import contextlib, decimal, statistics, random, json, atexit, time, array, ctypes, functools
+import contextlib, decimal, statistics, random, json, atexit, time, array, ctypes, functools, os, fcntl
 from tinygrad.helpers import PROFILEPATH, PROFILE, from_mv, getenv
 from tinygrad.renderer import Renderer
 from tinygrad.device import BufferSpec, Compiler, Compiled, LRUAllocator
+from tinygrad.runtime.autogen import libc
+
+# all HCQ interaction with the system happens through this Hardware Abstraction Layer. the devices should not make syscalls
+class HAL:
+  def __init__(self, path, flags): self.fd = os.open(path, flags)
+  def __del__(self): os.close(self.fd)
+  def ioctl(self, request, arg): return fcntl.ioctl(self.fd, request, arg)
+  def mmap(self, start, sz, prot, flags, offset): return libc.mmap(start, sz, prot, flags, self.fd, offset)
```

Comment:
I don't think flags should be an input here. Both ops_amd and ops_nv have lines defining constants